---
title: "scMultiMap"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette shows two examples of applying `scMultiMap` to infer peak-gene associations in cell types with single-cell multimodal data. 

```{r setup, warning = FALSE, message = FALSE}
library(scMultiMap)
library(Signac)
library(Seurat)
```

# 1. Toy example

To illustrate the basic functionality of scMultiMap, we use a toy Seurat object `small_obj` and a data frame of candidate peak-gene pairs `small_pairs_df` contained in the R package to run scMultiMap:
```{r}
# a Seurat object
small_obj

# candidate peak-gene pairs
head(small_pairs_df)
dim(small_pairs_df)
```

With these example data, we then run scMultiMap:
```{r}
res <- scMultiMap(small_obj, small_pairs_df) # < 0.1 seconds
head(res)
dim(res)
```

Here, scMultiMap takes a Seurat object with two modalities: RNA and peak, and tests the associations between 250 peak-gene pairs. We assume that this Seurat object contains the cells from a specific cell type of interest, such that the inferred associations are cell-type-specific.

# 2. PBMC example from 10x multiome data

# 2.1 Note: fragment counts are required for scMultiMap

Before demonstrating scMultiMap on real data, we make a note on the requirement for peak counts. As discussed in "Discussion" in the [scMultiMap manuscript](https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1), scMultiMap is designed to model **fragment counts**, which are amenable to count-based modeling and yield improved performance in downstream analysis compared to insertion counts or read counts ([Miao and Kim, 2024](https://www.nature.com/articles/s41592-023-02103-7), [Martens et al., 2024](https://www.nature.com/articles/s41592-023-02112-6)). Therefore, if you plan to use scMultiMap, please make sure the your scATAC-seq data are fragment counts instead of read counts. We recommend using MACS2 to call peaks which generates fragment counts, as shown in [Calling peaks](https://stuartlab.org/signac/articles/peak_calling). If you only have read counts, fragment counts can be estimated following [Martens et al., 2024](https://www.nature.com/articles/s41592-023-02112-6).

To demonstrate this, we downloaded a 10x multiome data (paired scRNA-seq and scATAC-seq) from 10x, used MACS2 to call peaks by cell types, and generated `processed_seurat_obj_PBMC_10k.rds`. The source code for these steps can be found at  [scMultiMap reproducibility Github repo](https://github.com/ChangSuBiostats/scMultiMap_analysis/preprocessing/preprocess_10x_PBMC_data.R).
<details>
<summary>Codes from downloading data from 10x</summary>

``` bash
wget https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5
wget https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz
wget https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi
```

</details>

## 2.2 Demo of scMultiMap on CD14 Monocytes

```{r eval = FALSE, include = TRUE}
pbmc <- readRDS('processed_seurat_obj_PBMC_10k.rds')
# in this example, we focus on CD14 monocytes:
pbmc_cd14 <- subset(pbmc, predicted.id == 'CD14 Mono')
print(pbmc_cd14)
```
```{r eval=TRUE, include=FALSE}
pbmc_cd14 <- readRDS('../../../data/CD14_obj.rds')
print(pbmc_cd14)
```

There are 3,146 CD14 monocytes in this dataset.

We choose to study the top $2{,}000$ highly expressed genes and top $20{,}000$ highly accessible peaks in CD14 Monocytes, and use a cis-region of width 1Mb to define candidate peak-gene pairs.


```{r eval=TRUE, include=TRUE}
pairs_df <- get_top_peak_gene_pairs(pbmc_cd14, 
                                    gene_top=2000, peak_top=20000, 
                                    distance = 5e+5,
                                    gene_assay = 'RNA', 
                                    peak_assay = 'peaks_ct') # name of the peak assay
head(pairs_df)
```

Here, we load 30,635 candidate peak-gene pairs generated based on the criterion above. `peak_gene_pairs` is a dataframe includes 30,635 rows for pairs and two columns for gene and peak names, respectively. Generally, these pairs should be formed and provided by the users.

Nect, we run `scMultiMap` to identify significantly associated peak-gene pairs

```{r eval=TRUE, include=TRUE}
res <- scMultiMap(pbmc_cd14, pairs_df, 
                  gene_assay = 'RNA',
                  peak_assay = 'peak')
head(res)
```

* Running scMultiMap on 3,146 cells for 30,635 peak-gene pairs will take a few minutes, depending on the machine.
* `res` is a dataframe with association statistics and p-values for each of the 30,635 peak-gene pairs.

# 2.3 Donwstream analysis with scMultiMap's results


**This vignettes in this package are currently under active development. More examples are coming soon!**
