<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scMultiMap for disease-control studies • scMultiMap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scMultiMap for disease-control studies">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMultiMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scMultiMap.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/scMultiMap.html">Introduction to scMultiMap</a></li>
    <li><a class="dropdown-item" href="../articles/disease_control.html">scMultiMap for disease-control studies</a></li>
    <li><a class="dropdown-item" href="../articles/GWAS.html">scMultiMap for integrative analysis with GWAS results</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav"></ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scMultiMap for disease-control studies</h1>
            
      

      <div class="d-none name"><code>disease_control.Rmd</code></div>
    </div>

    
    
<p>This vignette demonstrates the application of <code>scMultiMap</code>
to infer differentially associated peak-gene pairs using single-cell
multimodal data from disease-control studies.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://changsubiostats.github.io/scMultiMap/">scMultiMap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co"># for loading differentially expressed genes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ycphs.github.io/openxlsx/index.html" class="external-link">openxlsx</a></span><span class="op">)</span></span>
<span><span class="co"># for making heatmaps</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span></code></pre></div>
<p>In scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>,
we applied a permutation procedure to identify significantly
differentially associated peak-gene pairs between microglia from healthy
control subjects and subjects with Alzheimer’s disease (Results section
“scMultiMap mapped GWAS variants of Alzheimer’s disease to target genes
in microglia”). We illustrate the codes for this analysis here. At the
end of this vignette, we will generate Figures 4b from the
manuscript.</p>
<div class="section level2">
<h2 id="load-pre-processed-data">Load pre-processed data<a class="anchor" aria-label="anchor" href="#load-pre-processed-data"></a>
</h2>
<p>We downloaded the single-cell Multiome data on brain from an
Alzheimer’s disease study <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC10025452/" class="external-link">Anderson et
al.</a>, which is the same dataset we used in the manuscript (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE214979" class="external-link">GSE214979</a>).
We further re-called peaks by cell types with MACS2 using the codes at
the <a href="https://github.com/ChangSuBiostats/scMultiMap_analysis/preprocessing/preprocess_10x_PBMC_data.R" class="external-link">scMultiMap
reproducibility Github repo</a>. Calling peaks by cell types generates
cell-type-specific peaks that better represent candidate
cell-type-specific enhancers than those identified with all cell types
merged, which may obscure signals from less abundant cell types such as
microglia. Using MACS2 provides the fragment counts required for running
<code>scMultiMap</code> (see discussion <a href="scMultiMap.html">here</a>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set it to the directory where data are saved</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">'../../../data'</span> </span>
<span></span>
<span><span class="co"># load AD Multiome data</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s/processed_AD_DLPFC_15.rds'</span>, <span class="va">data_dir</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># subset to microglia</span></span>
<span><span class="va">ct</span> <span class="op">&lt;-</span> <span class="st">'Microglia'</span></span>
<span><span class="va">ct_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">obj</span>, subset <span class="op">=</span> <span class="va">predicted.id</span> <span class="op">==</span> <span class="va">ct</span><span class="op">)</span></span></code></pre></div>
<p>In this analysis, we study the top 5000 highly expressed genes and
top 50000 highly accessible peaks in microglia, and use a cis-region of
width 1Mb to define candidate peak-gene pairs.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_top_peak_gene_pairs.html">get_top_peak_gene_pairs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ct_obj</span>, subset <span class="op">=</span> <span class="va">Diagnosis</span> <span class="op">==</span> <span class="st">'Unaffected'</span><span class="op">)</span>,</span>
<span>                                    gene_top<span class="op">=</span><span class="fl">5000</span>, peak_top<span class="op">=</span><span class="fl">50000</span>,</span>
<span>                                    distance <span class="op">=</span> <span class="fl">5e+5</span>,</span>
<span>                                    gene_assay <span class="op">=</span> <span class="st">'RNA'</span>, <span class="co"># name of the gene assay</span></span>
<span>                                    peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span> <span class="co"># name of the peak assay</span></span>
<span><span class="co"># In the paper, we used control microglia to define peak-gene pairs. (`ct_control`)</span></span>
<span><span class="co"># One can also use all microglia.</span></span></code></pre></div>
<p>Here, we generated 135401 candidate peak-gene pairs based on the
criterion above. The users can also customize the peak-gene pairs based
on your analysis goal.</p>
</div>
<div class="section level2">
<h2 id="differential-association-analysis">Differential association analysis<a class="anchor" aria-label="anchor" href="#differential-association-analysis"></a>
</h2>
<p>Next, we estimate peak-gene associations in microglia from the
control and the Alzheimer’s disease groups, respectively. Note that
there are multiple biological samples in each group, so
<code>scMultiMap</code> should be run with <code>bsample</code> to avoid
spurious associations due to heterogeneity across biological samples
(see Methods in scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>
for more details).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># multiple biological samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ct_obj</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     1224     1230     1238     3329     3586     4305     4313     4443 </span></span>
<span><span class="co">#&gt;      201      161      283      211      152      347       17      171 </span></span>
<span><span class="co">#&gt;     4481     4482     4627 HCT17HEX   HCTZZT   NT1261   NT1271 </span></span>
<span><span class="co">#&gt;      233      415      241      264       26      192      265</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># control</span></span>
<span><span class="va">control_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ct_obj</span>, subset <span class="op">=</span> <span class="va">Diagnosis</span> <span class="op">==</span> <span class="st">'Unaffected'</span><span class="op">)</span>, </span>
<span>                          <span class="va">pairs_df</span>,</span>
<span>                          bsample <span class="op">=</span> <span class="st">'id'</span>, <span class="co"># labels for biological samples in the Seurat object</span></span>
<span>                          gene_assay <span class="op">=</span> <span class="st">'RNA'</span>, <span class="co"># name of the gene assay</span></span>
<span>                          peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span> <span class="co"># name of the peak assay</span></span>
<span><span class="co">#&gt; Start step 1: IRLS</span></span>
<span><span class="co">#&gt; Start IRLS for RNA</span></span>
<span><span class="co">#&gt; Start IRLS for peaks</span></span>
<span><span class="co">#&gt; Start step 2: WLS</span></span>
<span><span class="co">#&gt; There are 4652 unique genes in the peak-gene pairs.</span></span>
<span><span class="co">#&gt; Warning in sqrt(deno): NaNs produced</span></span>
<span><span class="co">#&gt; scMultiMap elapsed time: 01:11 (mm:ss)</span></span>
<span></span>
<span><span class="co"># disease</span></span>
<span><span class="va">disease_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ct_obj</span>, subset <span class="op">=</span> <span class="va">Diagnosis</span> <span class="op">==</span> <span class="st">"Alzheimer's"</span><span class="op">)</span>, </span>
<span>                          <span class="va">pairs_df</span>,</span>
<span>                          bsample <span class="op">=</span> <span class="st">'id'</span>,</span>
<span>                          gene_assay <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span>                          peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Start step 1: IRLS</span></span>
<span><span class="co">#&gt; Start IRLS for RNA</span></span>
<span><span class="co">#&gt; Start IRLS for peaks</span></span>
<span><span class="co">#&gt; Start step 2: WLS</span></span>
<span><span class="co">#&gt; There are 4652 unique genes in the peak-gene pairs.</span></span>
<span><span class="co">#&gt; scMultiMap elapsed time: 01:15 (mm:ss)</span></span></code></pre></div>
<p>The difference between two estimates (the ‘observed difference’)
represents the magnitude of differential association. To assess
statistical significance, we use a permutation test that randomly
reassigns sample labels to disease and control groups 100 times,
generating a null distribution of the difference between groups (the
‘null difference’).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtain the status for each id (sample)</span></span>
<span><span class="va">Status_id_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ct_obj</span><span class="op">$</span><span class="va">id</span>, <span class="va">ct_obj</span><span class="op">$</span><span class="va">Status</span><span class="op">)</span></span>
<span><span class="va">Status_id_tab</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;             AD Ctrl</span></span>
<span><span class="co">#&gt;   1224       0  201</span></span>
<span><span class="co">#&gt;   1230       0  161</span></span>
<span><span class="co">#&gt;   1238       0  283</span></span>
<span><span class="co">#&gt;   3329     211    0</span></span>
<span><span class="co">#&gt;   3586       0  152</span></span>
<span><span class="co">#&gt;   4305     347    0</span></span>
<span><span class="co">#&gt;   4313      17    0</span></span>
<span><span class="co">#&gt;   4443     171    0</span></span>
<span><span class="co">#&gt;   4481     233    0</span></span>
<span><span class="co">#&gt;   4482     415    0</span></span>
<span><span class="co">#&gt;   4627     241    0</span></span>
<span><span class="co">#&gt;   HCT17HEX   0  264</span></span>
<span><span class="co">#&gt;   HCTZZT     0   26</span></span>
<span><span class="co">#&gt;   NT1261     0  192</span></span>
<span><span class="co">#&gt;   NT1271     0  265</span></span>
<span><span class="va">Status_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Status_id_tab</span><span class="op">)</span><span class="op">[</span><span class="va">Status_id_tab</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_comb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Status_id_tab</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Status_ids</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">n_permu</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="co"># randomly reassign 7 samples to the AD group</span></span>
<span><span class="va">random_combs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">all_comb</span><span class="op">)</span>, <span class="va">n_permu</span>, replace <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>We then generated a permutation p-value by evaluating the proportion
of times that the “observed difference” is larger in magnitude than the
“null difference”.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diff_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="va">n_permu</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i_permu</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_permu</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># obtain randomly shuffled sample labels</span></span>
<span>  <span class="va">disease_sams</span> <span class="op">&lt;-</span> <span class="va">all_comb</span><span class="op">[</span>,<span class="va">random_combs</span><span class="op">[</span><span class="va">i_permu</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">control_sams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Status_id_tab</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Status_id_tab</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">disease_sams</span><span class="op">]</span></span>
<span>  <span class="co"># control</span></span>
<span>  <span class="va">null_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ct_obj</span>, subset <span class="op">=</span> <span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">control_sams</span><span class="op">)</span>, </span>
<span>                             <span class="va">pairs_df</span>,</span>
<span>                             bsample <span class="op">=</span> <span class="st">'id'</span>,</span>
<span>                             gene_assay <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span>                             peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># disease</span></span>
<span>  <span class="va">null_disease</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ct_obj</span>, subset <span class="op">=</span> <span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">disease_sams</span><span class="op">)</span>, </span>
<span>                             <span class="va">pairs_df</span>,</span>
<span>                             bsample <span class="op">=</span> <span class="st">'id'</span>,</span>
<span>                             gene_assay <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span>                             peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># null difference, i_permu</span></span>
<span>  <span class="va">diff_mat</span><span class="op">[</span>,<span class="va">i_permu</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">null_disease</span><span class="op">[</span>,<span class="st">'covar'</span><span class="op">]</span> <span class="op">-</span> <span class="va">null_control</span><span class="op">[</span>,<span class="st">'covar'</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="va">i_permu</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obs_diff</span> <span class="op">&lt;-</span> <span class="va">disease_res</span><span class="op">[</span>,<span class="st">'covar'</span><span class="op">]</span> <span class="op">-</span> <span class="va">control_res</span><span class="op">[</span>,<span class="st">'covar'</span><span class="op">]</span></span>
<span><span class="va">pval1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">obs_diff</span> <span class="op">&gt;</span> <span class="va">diff_mat</span><span class="op">)</span></span>
<span><span class="va">pval2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">obs_diff</span> <span class="op">&lt;</span> <span class="va">diff_mat</span><span class="op">)</span></span>
<span><span class="va">permu_pval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">pval1</span>, <span class="va">pval2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">2</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>cex <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">permu_pval</span>, main <span class="op">=</span> <span class="st">'Permutation p-values'</span>, breaks <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="./articles/figures/dc_pval.png" width="360" height="360" alt="Distribution of differential association p-values"><div class="figcaption">Distribution of differential association
p-values</div>
</div>
</div>
<div class="section level2">
<h2 id="downstream-analysis-of-differentially-associated-peak-gene-pairs">Downstream analysis of differentially associated peak-gene
pairs<a class="anchor" aria-label="anchor" href="#downstream-analysis-of-differentially-associated-peak-gene-pairs"></a>
</h2>
<div class="section level3">
<h3 id="enrichment-of-differentially-expressed-genes">Enrichment of differentially expressed genes<a class="anchor" aria-label="anchor" href="#enrichment-of-differentially-expressed-genes"></a>
</h3>
<p>We assess if genes with differential association are enriched for
differential expression.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># microglia DEG from Supplementary table 9 of https://www.nature.com/articles/s41586-024-07606-7</span></span>
<span><span class="va">DEG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s/aggregated_fullset.Mic_Immune_Mic.tsv.gz'</span>, <span class="va">data_dir</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">DEG</span> <span class="op">&lt;-</span> <span class="va">DEG</span><span class="op">[</span><span class="va">DEG</span><span class="op">$</span><span class="va">region</span> <span class="op">==</span> <span class="st">'PFC'</span>,<span class="op">]</span></span>
<span><span class="va">DEG</span> <span class="op">&lt;-</span> <span class="va">DEG</span><span class="op">[</span><span class="va">DEG</span><span class="op">$</span><span class="va">log10p_nm</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">DEG</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">DEG_inds</span> <span class="op">&lt;-</span> <span class="va">control_res</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">DEG</span><span class="op">$</span><span class="va">gene</span></span>
<span></span>
<span><span class="co"># evaluate enrichment of differentially expressed genes among genes from differentially associated peak-gene pairs</span></span>
<span><span class="va">total_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">DEG_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">DEG_inds</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sig_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">permu_pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span><span class="op">)</span> <span class="co"># genes in differentially associated peak-gene pairs (p-value &lt; 0.05)</span></span>
<span><span class="va">tab1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">total_gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">DEG_gene</span>, <span class="va">total_gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sig_gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html" class="external-link">fisher.test</a></span><span class="op">(</span><span class="va">tab1</span>, alternative <span class="op">=</span> <span class="st">'greater'</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Fisher's Exact Test for Count Data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data:  tab1</span></span>
<span><span class="co">#&gt; p-value = 6.035e-06</span></span>
<span><span class="co">#&gt; alternative hypothesis: true odds ratio is greater than 1</span></span>
<span><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span><span class="co">#&gt;  1.308737      Inf</span></span>
<span><span class="co">#&gt; sample estimates:</span></span>
<span><span class="co">#&gt; odds ratio </span></span>
<span><span class="co">#&gt;   1.556541</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># another set of microglia DEG from supplementary table of https://pmc.ncbi.nlm.nih.gov/articles/PMC10025452/</span></span>
<span><span class="co"># wget https://ars.els-cdn.com/content/image/1-s2.0-S2666979X23000198-mmc4.xls</span></span>
<span><span class="va">DEG</span> <span class="op">&lt;-</span> <span class="fu">read.xlsx</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s/1-s2.0-S2666979X23000198-mmc4.xlsx'</span>, <span class="va">data_dir</span><span class="op">)</span>, sheet <span class="op">=</span> <span class="st">'TableS3_ADCtrl_DEGs'</span><span class="op">)</span></span>
<span><span class="va">DEG_inds</span> <span class="op">&lt;-</span> <span class="va">control_res</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">DEG</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">DEG</span><span class="op">$</span><span class="va">celltype</span> <span class="op">==</span> <span class="st">'Microglia'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># evaluate enrichment of differentially expressed genes among genes from differentially associated peak-gene pairs</span></span>
<span><span class="va">total_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">DEG_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">DEG_inds</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sig_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">permu_pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span><span class="op">)</span> <span class="co"># genes in differentially associated peak-gene pairs (p-value &lt; 0.05)</span></span>
<span><span class="va">tab2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">total_gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">DEG_gene</span>, <span class="va">total_gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sig_gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html" class="external-link">fisher.test</a></span><span class="op">(</span><span class="va">tab2</span>, alternative <span class="op">=</span> <span class="st">'greater'</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Fisher's Exact Test for Count Data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data:  tab2</span></span>
<span><span class="co">#&gt; p-value = 0.01361</span></span>
<span><span class="co">#&gt; alternative hypothesis: true odds ratio is greater than 1</span></span>
<span><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span><span class="co">#&gt;  1.097925      Inf</span></span>
<span><span class="co">#&gt; sample estimates:</span></span>
<span><span class="co">#&gt; odds ratio </span></span>
<span><span class="co">#&gt;   1.482689</span></span></code></pre></div>
<p>Comparing with AD DEGs results from <a href="https://www.nature.com/articles/s41586-024-07606-7" class="external-link">Mathys et
al.</a> and <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC10025452/" class="external-link">Anderson et
al.</a>, the genes with differential assocations with peaks in microglia
are significantly enriched for microglia DEGs with odds ratio 1.56 and
1.48, respectively.</p>
</div>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p>We then visualize the associations in control versus disease groups.
We focus on peak-gene pairs that (1) are significantlly differentially
associated (nominal p-value &lt; 0.05); (2) are significantlly
associated in either control or AD microglia (p-value &lt; 0.05); (3)
have a difference of correlation greater than 0.2.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset_inds</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">permu_pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">|</span> <span class="va">disease_res</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">cor</span> <span class="op">-</span> <span class="va">disease_res</span><span class="op">$</span><span class="va">cor</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coexp_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">control_res</span><span class="op">$</span><span class="va">cor</span>, <span class="va">disease_res</span><span class="op">$</span><span class="va">cor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">coexp_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'control'</span>, <span class="st">'AD'</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">pheatmap</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">coexp_mat</span><span class="op">[</span><span class="va">DEG_inds</span> <span class="op">&amp;</span> <span class="va">subset_inds</span>,<span class="op">]</span><span class="op">)</span>,</span>
<span>            col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">11</span>, <span class="st">'RdBu'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            cluster_rows <span class="op">=</span> <span class="cn">F</span>, cluster_cols <span class="op">=</span> <span class="cn">T</span>,</span>
<span>            annotation_names_row <span class="op">=</span> <span class="cn">F</span>, </span>
<span>            annotation_names_col <span class="op">=</span> <span class="cn">F</span>,</span>
<span>            fontsize_row <span class="op">=</span> <span class="fl">15</span>,          <span class="co"># row label font size</span></span>
<span>            fontsize_col <span class="op">=</span> <span class="fl">7</span>,          <span class="co"># column label font size </span></span>
<span>            angle_col <span class="op">=</span> <span class="fl">45</span>, <span class="co"># sample names at an angle</span></span>
<span>            legend_breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0.2</span><span class="op">)</span>, <span class="co">#c(-1, 0, 1), # legend customisation</span></span>
<span>            show_colnames <span class="op">=</span> <span class="cn">F</span>, show_rownames <span class="op">=</span> <span class="cn">T</span>, <span class="co"># displaying column and row names</span></span>
<span>            cellheight<span class="op">=</span><span class="fl">50</span>, treeheight_col <span class="op">=</span> <span class="fl">15</span>,</span>
<span>            fontsize <span class="op">=</span> <span class="fl">12</span><span class="op">)</span> </span></code></pre></div>
<div class="float">
<img src="./articles/figures/dc_heatmap.png" style="width:65.0%" alt="Heatmap of differential peak-gene association"><div class="figcaption">Heatmap of differential peak-gene
association</div>
</div>
<p>This reproduced Figure~4b in the scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>.</p>
<p>These differentially associated peak-gene pairs can be further
analyzed with GWAS (Genome-wide association studies) results, to
elucidate the cell-type-specific regulatory mechanisms of GWAS variant
in disease relevant cell types. Please see <a href="GWAS.html">scMultiMap for integrative analysis with GWAS
results</a>, a separate vignette that uses differentially associated
peak-gene pairs identified here to investigate the function of causal
variants of Alzheimer’s disease in microglia.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
