<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scMultiMap for integrative analysis with GWAS results • scMultiMap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scMultiMap for integrative analysis with GWAS results">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMultiMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scMultiMap.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav"></ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scMultiMap for integrative analysis with GWAS results</h1>
            
      

      <div class="d-none name"><code>GWAS.Rmd</code></div>
    </div>

    
    
<p>This vignette demonstrates how to apply scMultiMap to map associated
peak-gene pairs and integrate the results with genome-wide association
studies (GWAS) to elucidate the regulatory functions of GWAS variants in
disease-associated cell types. As an example, we will reproduce the
results from the <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">scMultiMap
manuscript</a>, specifically those presented in the section ‘scMultiMap
mapped GWAS variants of Alzheimer’s disease to target genes in
microglia.’ At the end of this vignette, we will generate Figures 4d and
4e from the manuscript.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://changsubiostats.github.io/scMultiMap/">scMultiMap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ycphs.github.io/openxlsx/index.html" class="external-link">openxlsx</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h2>
<p>Over 75 genetic loci have been associated with the risk of developing
Alzheimer’s disease (AD) (<a href="https://www.nature.com/articles/s41588-022-01024-z" class="external-link">Bellenguez, C.
et al.</a>), yet their regulatory functions and relevant cellular
contexts remain unclear. Among brain cell types, microglia—the brain’s
innate immune cells—exhibit the highest enrichment of genetic
association with AD in their cis-regulatory elements (<a href="https://www.science.org/doi/10.1126/science.aay0793" class="external-link">Nott, A. et
al.</a>, <a href="https://www.nature.com/articles/nature14252" class="external-link">Gjoneska,
E. et al.</a>). Therefore, this analysis focuses on microglia and uses
<code>scMultiMap</code> to identify candidate pairs of cis-regulatory
elements and their target genes. As two examples, we examine how
<code>scMultiMap</code> results can elucidate the regulatory function of
fine-mapped AD variants in microglia at two loci: PICALM and INPP5D.</p>
</div>
<div class="section level2">
<h2 id="run-scmultimap">Run scMultiMap<a class="anchor" aria-label="anchor" href="#run-scmultimap"></a>
</h2>
<p>Note: this vignette use the same dataset as in <a href="disease_control.html">vignette: scMultiMap for disease-control
studies</a>. Please refer to it for details on pre-processing data and
running <code>scMultiMap</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set it to the directory where data are saved</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">'../../../data'</span> </span>
<span></span>
<span><span class="co"># load AD Multiome data</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s/processed_AD_DLPFC_15.rds'</span>, <span class="va">data_dir</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># subset to microglia</span></span>
<span><span class="va">ct</span> <span class="op">&lt;-</span> <span class="st">'Microglia'</span></span>
<span><span class="va">ct_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">obj</span>, subset <span class="op">=</span> <span class="va">predicted.id</span> <span class="op">==</span> <span class="va">ct</span><span class="op">)</span></span></code></pre></div>
<p>In this analysis, we study the top 5000 highly expressed genes and
top 50000 highly accessible peaks in microglia, and use a cis-region of
width 1Mb to define candidate peak-gene pairs.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_top_peak_gene_pairs.html">get_top_peak_gene_pairs</a></span><span class="op">(</span><span class="va">ct_control</span>,</span>
<span>                                    gene_top<span class="op">=</span><span class="fl">5000</span>, peak_top<span class="op">=</span><span class="fl">50000</span>,</span>
<span>                                    distance <span class="op">=</span> <span class="fl">5e+5</span>,</span>
<span>                                    gene_assay <span class="op">=</span> <span class="st">'RNA'</span>, <span class="co"># name of the gene assay</span></span>
<span>                                    peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span> <span class="co"># name of the peak assay</span></span>
<span><span class="co"># In the paper, we used control microglia to define peak-gene pairs. (`ct_control`)</span></span>
<span><span class="co"># One can also use all microglia.</span></span></code></pre></div>
<p>Here, we generated 135401 candidate peak-gene pairs based on the
criterion above. The users can also customize the peak-gene pairs based
on your analysis goal.</p>
<p>Next, we estimate peak-gene associations in microglia from the
control and the Alzheimer’s disease groups, respectively. Note that
there are multiple biological samples in this dataset, so
<code>scMultiMap</code> should be run with <code>bsample</code> to avoid
spurious associations due to heterogeneity across biological samples
(see Methods in scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>
for more details).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># multiple biological samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ct_obj</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     1224     1230     1238     3329     3586     4305     4313     4443 </span></span>
<span><span class="co">#&gt;      201      161      283      211      152      347       17      171 </span></span>
<span><span class="co">#&gt;     4481     4482     4627 HCT17HEX   HCTZZT   NT1261   NT1271 </span></span>
<span><span class="co">#&gt;      233      415      241      264       26      192      265</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># control</span></span>
<span><span class="va">control_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="va">ct_control</span>, </span>
<span>                          <span class="va">pairs_df</span>,</span>
<span>                          bsample <span class="op">=</span> <span class="st">'id'</span>,</span>
<span>                          gene_assay <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span>                          peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Start step 1: IRLS</span></span>
<span><span class="co">#&gt; Start IRLS for RNA</span></span>
<span><span class="co">#&gt; Start IRLS for peaks</span></span>
<span><span class="co">#&gt; Start step 2: WLS</span></span>
<span><span class="co">#&gt; There are 4652 unique genes in the peak-gene pairs.</span></span>
<span><span class="co">#&gt; Warning in sqrt(deno): NaNs produced</span></span>
<span><span class="co">#&gt; scMultiMap elapsed time: 01:13 (mm:ss)</span></span>
<span></span>
<span><span class="co"># disease</span></span>
<span><span class="va">disease_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="va">ct_AD</span>, </span>
<span>                          <span class="va">pairs_df</span>,</span>
<span>                          bsample <span class="op">=</span> <span class="st">'id'</span>,</span>
<span>                          gene_assay <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span>                          peak_assay <span class="op">=</span> <span class="st">'peaks'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Start step 1: IRLS</span></span>
<span><span class="co">#&gt; Start IRLS for RNA</span></span>
<span><span class="co">#&gt; Start IRLS for peaks</span></span>
<span><span class="co">#&gt; Start step 2: WLS</span></span>
<span><span class="co">#&gt; There are 4652 unique genes in the peak-gene pairs.</span></span>
<span><span class="co">#&gt; scMultiMap elapsed time: 01:16 (mm:ss)</span></span></code></pre></div>
<details><summary>
What does “Warning in sqrt(deno): NaNs produced” mean?
</summary><p>This warning suggests that some peaks or genes have variance
estimated to be negative, causing numerical errors in calculating
correlations. This situation is typically rare and may happen when you
consider genes or peaks with extremely low abundance (e.g. top 5000
genes and top 50000 peaks here). The estimated correlations will be
internally set to 0 by <code>scMultiMap</code>.</p>
</details>
</div>
<div class="section level2">
<h2 id="load-results-on-fine-mapped-ad-gwas-variants">Load results on fine-mapped AD GWAS variants<a class="anchor" aria-label="anchor" href="#load-results-on-fine-mapped-ad-gwas-variants"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># AD GWAS variants fine-mapped and priortized based on eQTL &amp; Hi-C</span></span>
<span><span class="co"># "Seeding from PAINTOR or FGWAS 95% credible set variants (Sup- plementary Table 6b), </span></span>
<span><span class="co"># we first included their linkage disequilibrium (LD) tags (defined as LD R2 &gt; 0.8) in TOP-LD European (EUR) </span></span>
<span><span class="co"># and then retained only those overlapping microglia ATAC–seq peaks that interact with promoters"</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># Reference: https://www.nature.com/articles/s41588-023-01506-8</span></span>
<span><span class="co"># Data source: supplementary table 6b</span></span>
<span></span>
<span><span class="va">mic_fm_prior_variants_df</span> <span class="op">&lt;-</span> <span class="fu">read.xlsx</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s/41588_2023_1506_MOESM3_ESM.xlsx'</span>, <span class="va">data_dir</span><span class="op">)</span>,</span>
<span>                                sheet <span class="op">=</span> <span class="st">'Supplementary Table 6c'</span>, startRow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">mic_fm_prior_variants_df</span><span class="op">$</span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'chr'</span>, <span class="va">mic_fm_prior_variants_df</span><span class="op">$</span><span class="va">chr</span><span class="op">)</span></span>
<span><span class="va">grange_mic_fm_prior</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">mic_fm_prior_variants_df</span>, </span>
<span>                                                         seqnames.field <span class="op">=</span> <span class="st">'chr'</span>,</span>
<span>                                                         start.field <span class="op">=</span> <span class="st">'pos.hg38'</span>,</span>
<span>                                                         end.field <span class="op">=</span> <span class="st">'pos.hg38'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-inferred-peak-gene-links-and-fine-mapped-gwas-variants">Visualize inferred peak-gene links and fine-mapped GWAS
variants<a class="anchor" aria-label="anchor" href="#visualize-inferred-peak-gene-links-and-fine-mapped-gwas-variants"></a>
</h2>
<div class="section level3">
<h3 id="picalm-locus-figure-4d">PICALM locus (figure 4d)<a class="anchor" aria-label="anchor" href="#picalm-locus-figure-4d"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make grange objects that connect genes with peaks</span></span>
<span><span class="va">gene_peak_link_gr_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_gene_to_peak_link.html">make_gene_to_peak_link</a></span><span class="op">(</span><span class="va">control_res</span>, <span class="st">'PICALM'</span>, <span class="va">ct_control</span>, <span class="st">'peaks'</span><span class="op">)</span></span>
<span><span class="va">gene_peak_link_gr_AD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_gene_to_peak_link.html">make_gene_to_peak_link</a></span><span class="op">(</span><span class="va">disease_res</span>, <span class="st">'PICALM'</span>, <span class="va">ct_AD</span>, <span class="st">'peaks'</span><span class="op">)</span></span>
<span><span class="va">gene_peak_link_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>AD <span class="op">=</span> <span class="va">gene_peak_link_gr_AD</span>, control <span class="op">=</span> <span class="va">gene_peak_link_gr_control</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uniform_fs</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="co"># genomic region around PICALM locus</span></span>
<span><span class="va">selected_range</span> <span class="op">&lt;-</span> <span class="st">"chr11-85850000-86216290"</span></span>
<span></span>
<span><span class="va">gene_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/AnnotationPlot.html" class="external-link">AnnotationPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">ct_control</span>,</span>
<span>  region <span class="op">=</span> <span class="va">selected_range</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">link_plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">g</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'control'</span>,<span class="st">'AD'</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://stuartlab.org/signac/reference/Links.html" class="external-link">Links</a></span><span class="op">(</span><span class="va">ct_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_peak_link_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">gene_peak_link_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">score</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.2</span><span class="op">]</span></span>
<span>    <span class="va">link_plot_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/LinkPlot.html" class="external-link">LinkPlot</a></span><span class="op">(</span></span>
<span>        object <span class="op">=</span> <span class="va">ct_obj</span>,</span>
<span>        region <span class="op">=</span> <span class="va">selected_range</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"blue"</span>,</span>
<span>                              midpoint <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                              mid <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>                              high <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                              space<span class="op">=</span><span class="st">"Lab"</span>,</span>
<span>                             limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">'correlation'</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s'</span>, <span class="va">g</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expr_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/ExpressionPlot.html" class="external-link">ExpressionPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">ct_obj</span>,</span>
<span>  features <span class="op">=</span> <span class="st">"PICALM"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>    group.by<span class="op">=</span><span class="st">'Status'</span></span>
<span><span class="op">)</span>  <span class="op">+</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to see it on the plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">grange_mic_fm_prior</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">grange_mic_fm_prior</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1000</span></span>
<span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">ct_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'peaks'</span></span>
<span><span class="va">snp_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/PeakPlot.html" class="external-link">PeakPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">ct_obj</span>,</span>
<span>  <span class="va">selected_range</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peaks <span class="op">=</span> <span class="va">grange_mic_fm_prior</span>,</span>
<span>  group.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">'SNPs'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CombineTracks.html" class="external-link">CombineTracks</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">gene_plot</span>, <span class="va">snp_plot</span>, <span class="va">link_plot_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">link_plot_list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                   expression.plot <span class="op">=</span> <span class="va">expr_plot</span>,</span>
<span>                   heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">5</span><span class="op">/</span><span class="fl">7</span>,</span>
<span>                   widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6.5</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre></div>
<p><img src="figures/PICALM_links.png" style="width:65.0%"></p>
<p>This reproduced Figure~4d in the scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>.</p>
</div>
<div class="section level3">
<h3 id="inpp5d-locus-figure-4e">INPP5D locus (figure 4e)<a class="anchor" aria-label="anchor" href="#inpp5d-locus-figure-4e"></a>
</h3>
<p>Repeat the analysis above for the INPP5D locus.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make grange objects that connect genes with peaks</span></span>
<span><span class="va">gene_peak_link_gr_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_gene_to_peak_link.html">make_gene_to_peak_link</a></span><span class="op">(</span><span class="va">control_res</span>, <span class="st">'INPP5D'</span>, <span class="va">ct_control</span>, <span class="st">'peaks'</span><span class="op">)</span></span>
<span><span class="va">gene_peak_link_gr_AD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_gene_to_peak_link.html">make_gene_to_peak_link</a></span><span class="op">(</span><span class="va">disease_res</span>, <span class="st">'INPP5D'</span>, <span class="va">ct_AD</span>, <span class="st">'peaks'</span><span class="op">)</span></span>
<span><span class="va">gene_peak_link_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>AD <span class="op">=</span> <span class="va">gene_peak_link_gr_AD</span>, control <span class="op">=</span> <span class="va">gene_peak_link_gr_control</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uniform_fs</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="co"># genomic region around INPP5D locus</span></span>
<span><span class="va">selected_range</span> <span class="op">&lt;-</span> <span class="st">"chr2-232850000-233259353"</span></span>
<span></span>
<span><span class="va">gene_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/AnnotationPlot.html" class="external-link">AnnotationPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">ct_control</span>,</span>
<span>  region <span class="op">=</span> <span class="va">selected_range</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">link_plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">g</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'control'</span>,<span class="st">'AD'</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://stuartlab.org/signac/reference/Links.html" class="external-link">Links</a></span><span class="op">(</span><span class="va">ct_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_peak_link_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">gene_peak_link_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">score</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.2</span><span class="op">]</span></span>
<span>    <span class="va">link_plot_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/LinkPlot.html" class="external-link">LinkPlot</a></span><span class="op">(</span></span>
<span>        object <span class="op">=</span> <span class="va">ct_obj</span>,</span>
<span>        region <span class="op">=</span> <span class="va">selected_range</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"blue"</span>,</span>
<span>                              midpoint <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                              mid <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>                              high <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                              space<span class="op">=</span><span class="st">"Lab"</span>,</span>
<span>                             limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">'correlation'</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s'</span>, <span class="va">g</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expr_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/ExpressionPlot.html" class="external-link">ExpressionPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">ct_obj</span>,</span>
<span>  features <span class="op">=</span> <span class="st">"INPP5D"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>    group.by<span class="op">=</span><span class="st">'Status'</span></span>
<span><span class="op">)</span>  <span class="op">+</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to see it on the plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">grange_mic_fm_prior</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">grange_mic_fm_prior</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1000</span></span>
<span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">ct_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'peaks'</span></span>
<span><span class="va">snp_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/PeakPlot.html" class="external-link">PeakPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">ct_obj</span>,</span>
<span>  <span class="va">selected_range</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peaks <span class="op">=</span> <span class="va">grange_mic_fm_prior</span>,</span>
<span>  group.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">'SNPs'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">uniform_fs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CombineTracks.html" class="external-link">CombineTracks</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">gene_plot</span>, <span class="va">snp_plot</span>, <span class="va">link_plot_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">link_plot_list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                   expression.plot <span class="op">=</span> <span class="va">expr_plot</span>,</span>
<span>                   heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">5</span><span class="op">/</span><span class="fl">7</span>,</span>
<span>                   widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6.5</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre></div>
<p><img src="figures/INPP5D_links.png" style="width:65.0%"></p>
<p>This reproduced Figure~4e in the scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
