<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to scMultiMap • scMultiMap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to scMultiMap">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMultiMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/scMultiMap.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/scMultiMap.html">Introduction to scMultiMap</a></li>
    <li><a class="dropdown-item" href="../articles/disease_control.html">scMultiMap for disease-control studies</a></li>
    <li><a class="dropdown-item" href="../articles/GWAS.html">scMultiMap for integrative analysis with GWAS results</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav"></ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to scMultiMap</h1>
            
      

      <div class="d-none name"><code>scMultiMap.Rmd</code></div>
    </div>

    
    
<p>This vignette shows two examples of applying <code>scMultiMap</code>
to infer cell-type-specific peak-gene associations with single-cell
multimodal data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://changsubiostats.github.io/scMultiMap/">scMultiMap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co"># for consistency analysis </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="toy-example">1. Toy example<a class="anchor" aria-label="anchor" href="#toy-example"></a>
</h2>
<p>To illustrate the basic functionality of <code>scMultiMap</code>, we
use a toy Seurat object <code>small_obj</code> and a data frame of
candidate peak-gene pairs <code>small_pairs_df</code> provided in the R
package to run <code>scMultiMap</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a Seurat object</span></span>
<span><span class="va">small_obj</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 350 features across 100 samples within 2 assays </span></span>
<span><span class="co">#&gt; Active assay: RNA (100 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt;  1 other assay present: peak</span></span>
<span></span>
<span><span class="co"># candidate peak-gene pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">small_pairs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;     gene                    peak</span></span>
<span><span class="co">#&gt; 1 MALAT1 chr11-65013281-65014841</span></span>
<span><span class="co">#&gt; 2 MALAT1 chr11-65026934-65027962</span></span>
<span><span class="co">#&gt; 3 MALAT1 chr11-65040583-65041780</span></span>
<span><span class="co">#&gt; 4 MALAT1 chr11-65083457-65084665</span></span>
<span><span class="co">#&gt; 5 MALAT1 chr11-65095679-65096871</span></span>
<span><span class="co">#&gt; 6 MALAT1 chr11-65110172-65110785</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">small_pairs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 250   2</span></span></code></pre></div>
<p>With these example data, we then run <code>scMultiMap</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="va">small_obj</span>, <span class="va">small_pairs_df</span><span class="op">)</span> <span class="co"># &lt; 0.1 seconds</span></span>
<span><span class="co">#&gt; Start step 1: IRLS</span></span>
<span><span class="co">#&gt; Start IRLS for RNA</span></span>
<span><span class="co">#&gt; Start IRLS for peak</span></span>
<span><span class="co">#&gt; Start step 2: WLS</span></span>
<span><span class="co">#&gt; There are 10 unique genes in the peak-gene pairs.</span></span>
<span><span class="co">#&gt; scMultiMap elapsed time: 00:00 (mm:ss)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt;     gene                    peak       pval  test_stat         covar      padj</span></span>
<span><span class="co">#&gt; 1 MALAT1 chr11-65013281-65014841 0.15215553  1.4319591  3.100282e-05 0.8273097</span></span>
<span><span class="co">#&gt; 2 MALAT1 chr11-65026934-65027962 0.79384145  0.2613256  5.204042e-06 0.9753992</span></span>
<span><span class="co">#&gt; 3 MALAT1 chr11-65040583-65041780 0.68116942  0.4108679  8.623731e-06 0.9472305</span></span>
<span><span class="co">#&gt; 4 MALAT1 chr11-65083457-65084665 0.03236884  2.1398259  4.879823e-05 0.7196890</span></span>
<span><span class="co">#&gt; 5 MALAT1 chr11-65095679-65096871 0.02264609 -2.2793530 -4.308216e-05 0.7196890</span></span>
<span><span class="co">#&gt; 6 MALAT1 chr11-65110172-65110785 0.88810490 -0.1407026 -1.919029e-06 0.9881934</span></span>
<span><span class="co">#&gt;           cor</span></span>
<span><span class="co">#&gt; 1  0.00000000</span></span>
<span><span class="co">#&gt; 2  0.10216095</span></span>
<span><span class="co">#&gt; 3  0.29533510</span></span>
<span><span class="co">#&gt; 4  0.62091902</span></span>
<span><span class="co">#&gt; 5 -0.57664094</span></span>
<span><span class="co">#&gt; 6 -0.08605519</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 250   7</span></span></code></pre></div>
<p>Here, scMultiMap takes a Seurat object with two modalities: RNA and
peak, and evaluates the associations between 250 peak-gene pairs. We
assume that this Seurat object contains the cells from a specific cell
type of interest, such that the inferred associations are
cell-type-specific.</p>
</div>
<div class="section level2">
<h2 id="pbmc-example-10x-multiome-data">2. PBMC example, 10x Multiome data<a class="anchor" aria-label="anchor" href="#pbmc-example-10x-multiome-data"></a>
</h2>
<div class="section level3">
<h3 id="note-fragment-counts-are-required-for-scmultimap">2.1 Note: fragment counts are required for scMultiMap<a class="anchor" aria-label="anchor" href="#note-fragment-counts-are-required-for-scmultimap"></a>
</h3>
<p>Before demonstrating <code>scMultiMap</code> on real data, we
highlight an important requirement for peak counts. As discussed in
“Discussion” in scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>,
scMultiMap is designed to model <strong>fragment counts</strong>, which
are more suitable for count-based modeling and improve downstream
analysis performance compared to insertion counts or read counts (<a href="https://www.nature.com/articles/s41592-023-02103-7" class="external-link">Miao and Kim,
2024</a>, <a href="https://www.nature.com/articles/s41592-023-02112-6" class="external-link">Martens et
al., 2024</a>). Therefore, if you plan to use <code>scMultiMap</code>,
ensure the your scATAC-seq data consist of fragment counts instead of
read counts. We recommend using MACS2 for peak calling, as it generates
fragment counts (see <a href="https://stuartlab.org/signac/articles/peak_calling" class="external-link">Calling
peaks</a>). If you only have read counts and do not have access to the
ATAC-seq fragment files, fragment counts can be estimated following the
approach in <a href="https://www.nature.com/articles/s41592-023-02112-6" class="external-link">Martens et
al., 2024</a>.</p>
<p>To demonstrate this, we downloaded a 10x Multiome dataset (paired
scRNA-seq and scATAC-seq) from 10x, used MACS2 to call peaks by cell
types, and generated <code>processed_seurat_obj_PBMC_10k.rds</code>. The
source code for these steps can be found at <a href="https://github.com/ChangSuBiostats/scMultiMap_analysis/preprocessing/preprocess_10x_PBMC_data.R" class="external-link">scMultiMap
reproducibility Github repo</a>. Calling peak by cell type is to
generate cell-type-specific peaks that better represent candidate
cell-type-specific enhancers than those identified from merged cell
types.</p>
</div>
<div class="section level3">
<h3 id="demo-of-scmultimap-on-cd14-monocytes">2.2 Demo of scMultiMap on CD14 Monocytes<a class="anchor" aria-label="anchor" href="#demo-of-scmultimap-on-cd14-monocytes"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set it to the directory where data are saved</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">'../../../data'</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s/processed_seurat_obj_PBMC_10k_nextgem.rds'</span>, <span class="va">data_dir</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># in this example, we focus on CD14 monocytes:</span></span>
<span><span class="va">pbmc_cd14</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">predicted.id</span> <span class="op">==</span> <span class="st">'CD14 Mono'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pbmc_cd14</span><span class="op">)</span></span></code></pre></div>
<p>We choose to study the top 2000 highly expressed genes and top 20000
highly accessible peaks in CD14 Monocytes, and use a cis-region of width
1Mb to define candidate peak-gene pairs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pairs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_top_peak_gene_pairs.html">get_top_peak_gene_pairs</a></span><span class="op">(</span><span class="va">pbmc_cd14</span>,</span>
<span>                                    gene_top<span class="op">=</span><span class="fl">2000</span>, peak_top<span class="op">=</span><span class="fl">20000</span>,</span>
<span>                                    distance <span class="op">=</span> <span class="fl">5e+5</span>,</span>
<span>                                    gene_assay <span class="op">=</span> <span class="st">'RNA'</span>, <span class="co"># name of the gene assay</span></span>
<span>                                    peak_assay <span class="op">=</span> <span class="st">'peaks_ct'</span><span class="op">)</span> <span class="co"># name of the peak assay</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pairs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;     gene                    peak</span></span>
<span><span class="co">#&gt; 1 MALAT1 chr11-65614961-65616132</span></span>
<span><span class="co">#&gt; 2 MALAT1 chr11-65918502-65920142</span></span>
<span><span class="co">#&gt; 3 MALAT1 chr11-65856895-65858482</span></span>
<span><span class="co">#&gt; 4 MALAT1 chr11-65496994-65497909</span></span>
<span><span class="co">#&gt; 5 MALAT1 chr11-65569536-65570572</span></span>
<span><span class="co">#&gt; 6 MALAT1 chr11-65711306-65712797</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pairs_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 32296     2</span></span></code></pre></div>
<p>Here, we generated 32296 candidate peak-gene pairs generated based on
the criterion above. Generally, the pairs to be studied could also be
customized and supplied by users.</p>
<p>Nect, we run <code>scMultiMap</code> to assess peak-gene
associations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMultiMap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMultiMap.html">scMultiMap</a></span><span class="op">(</span><span class="va">pbmc_cd14</span>, <span class="va">pairs_df</span>,</span>
<span>                  gene_assay <span class="op">=</span> <span class="st">'RNA'</span>, <span class="co"># name of the gene assay</span></span>
<span>                  peak_assay <span class="op">=</span> <span class="st">'peaks_ct'</span><span class="op">)</span> <span class="co"># name of the peak assay</span></span>
<span><span class="co">#&gt; Start step 1: IRLS</span></span>
<span><span class="co">#&gt; Start IRLS for RNA</span></span>
<span><span class="co">#&gt; Start IRLS for peaks_ct</span></span>
<span><span class="co">#&gt; Start step 2: WLS</span></span>
<span><span class="co">#&gt; There are 1909 unique genes in the peak-gene pairs.</span></span>
<span><span class="co">#&gt; scMultiMap elapsed time: 00:19 (mm:ss)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMultiMap_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;     gene                    peak       pval test_stat        covar      padj</span></span>
<span><span class="co">#&gt; 1 MALAT1 chr11-65614961-65616132 0.90652298 0.1174254 1.743388e-09 0.9836402</span></span>
<span><span class="co">#&gt; 2 MALAT1 chr11-65918502-65920142 0.03734129 2.0820129 3.005705e-08 0.3367015</span></span>
<span><span class="co">#&gt; 3 MALAT1 chr11-65856895-65858482 0.26443486 1.1159703 1.604796e-08 0.7232544</span></span>
<span><span class="co">#&gt; 4 MALAT1 chr11-65496994-65497909 0.08524652 1.7210237 2.334859e-08 0.4901409</span></span>
<span><span class="co">#&gt; 5 MALAT1 chr11-65569536-65570572 0.18919341 1.3129690 1.776141e-08 0.6525889</span></span>
<span><span class="co">#&gt; 6 MALAT1 chr11-65711306-65712797 0.25426633 1.1400479 1.496624e-08 0.7139586</span></span>
<span><span class="co">#&gt;          cor</span></span>
<span><span class="co">#&gt; 1 0.01180729</span></span>
<span><span class="co">#&gt; 2 0.16766414</span></span>
<span><span class="co">#&gt; 3 0.07195509</span></span>
<span><span class="co">#&gt; 4 0.15467261</span></span>
<span><span class="co">#&gt; 5 0.09259916</span></span>
<span><span class="co">#&gt; 6 0.09048044</span></span></code></pre></div>
<ul>
<li>Running <code>scMultiMap</code> on this dataset took ~20-25 seconds
using a single core of an Intel(R) Xeon(R) Gold 6326 CPU @ 2.90GHz.
.</li>
<li>
<code>res</code> is a dataframe with association statistics and
p-values for each of 32296 peak-gene pairs.</li>
</ul>
</div>
<div class="section level3">
<h3 id="downstream-analysis-consistency-analysis">2.3 Downstream analysis: consistency analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis-consistency-analysis"></a>
</h3>
<p>To demonstrate <code>scMultiMap</code> is powerful for identifying
cell-type-specific enhancer-gene pairs, we consider the examples
presented in Figure~2b of the <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>
and evaluate the consistency (overlap) between results from
<code>scMultiMap</code> and that from promoter capture Hi-C (PCHiC).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Data required for liftover form hg19 to hg38</span></span>
<span><span class="co"># from http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz</span></span>
<span><span class="va">chain</span> <span class="op">=</span> <span class="fu">import.chain</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/hg19ToHg38.over.chain"</span>, <span class="va">data_dir</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>First, we preprocess PCHiC data and lift it over from hg19 to hg38 to
match the genome version used in the 10x Multiome data (hg38).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Obtain promoter capture Hi-C data from http://dx.doi.org/10.1016/j.cell.2016.09.037.</span></span>
<span><span class="co"># supplementary data "DATA_S1/PCHiC_peak_matrix_cutoff5.tsv"</span></span>
<span></span>
<span><span class="va">PCHiC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">'%s/PCHiC_peak_matrix_cutoff5.tsv'</span>, <span class="va">data_dir</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co"># focus on CD14 monocytes as denoted by the column 'Mon'</span></span>
<span><span class="va">PCHiC</span> <span class="op">&lt;-</span> <span class="va">PCHiC</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'oeChr'</span>, <span class="st">'oeStart'</span>, <span class="st">'oeEnd'</span>, <span class="co"># peak info</span></span>
<span>                  <span class="st">'baitName'</span>, <span class="co"># gene name,</span></span>
<span>                  <span class="st">'Mon'</span><span class="op">)</span><span class="op">]</span> <span class="co"># CHiCAGO scores for CD14 monocytes</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">PCHiC</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>, <span class="st">'start'</span>, <span class="st">'end'</span><span class="op">)</span></span>
<span><span class="va">grange_hg19</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">PCHiC</span>, keep.extra.columns <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">grange_hg19</span><span class="op">)</span> <span class="op">=</span> <span class="st">"UCSC"</span></span>
<span><span class="va">grange_hg38</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">liftOver</span><span class="op">(</span><span class="va">grange_hg19</span>, <span class="va">chain</span><span class="op">)</span></span>
<span><span class="va">grange_hg38</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">grange_hg38</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">grange_hg38</span><span class="op">)</span> <span class="op">=</span> <span class="st">"hg38"</span></span>
<span></span>
<span><span class="co"># significantly associated peak-gene pairs</span></span>
<span><span class="co"># defined by CHiCAGO scores &gt;=5 (reference: http://dx.doi.org/10.1016/j.cell.2016.09.037)</span></span>
<span><span class="va">sig_assay_gr</span> <span class="op">&lt;-</span> <span class="va">grange_hg38</span><span class="op">[</span><span class="va">grange_hg38</span><span class="op">@</span><span class="va">elementMetadata</span><span class="op">[[</span><span class="st">'Mon'</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">sig_assay_gr</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">sig_assay_gr</span><span class="op">$</span><span class="va">baitName</span></span></code></pre></div>
<div class="section level4">
<h4 id="scmultimap">scMultiMap<a class="anchor" aria-label="anchor" href="#scmultimap"></a>
</h4>
<p>We then evaluate its overlap with <code>scMultiMap</code>’s
results.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMultiMap_overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/validate_with_assay.html">validate_with_assay</a></span><span class="op">(</span><span class="va">scMultiMap_res</span>, <span class="va">sig_assay_gr</span>, pvar <span class="op">=</span> <span class="st">"padj"</span>, p_cutoff <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "p value from one-sided Fisher exact test: 5.39e-12, #overlap: 246, odds ratio: 1.74"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="signac">Signac<a class="anchor" aria-label="anchor" href="#signac"></a>
</h4>
<p>In comparison, we also evaluated <code>Signac</code>’s results.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By default, Signac uses normalized data to compute peak-gene associations</span></span>
<span><span class="va">pbmc_cd14</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_cd14</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##### </span></span>
<span><span class="co">##### The code below could take 3 hours to run, as</span></span>
<span><span class="co">#####   (1) it consider all peaks instead of peaks with high abundance</span></span>
<span><span class="co">#####   (2) constructing the sampling-based null distribution is computationally intensive</span></span>
<span><span class="co"># Run Signac</span></span>
<span><span class="va">signac_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/LinkPeaks.html" class="external-link">LinkPeaks</a></span><span class="op">(</span><span class="va">pbmc_cd14</span>, <span class="st">'peaks_ct'</span>, <span class="st">'RNA'</span>,</span>
<span>                        genes.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pairs_df</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>, <span class="co"># consider the same genes as above</span></span>
<span>                        pvalue_cutoff <span class="op">=</span> <span class="fl">1</span>, score_cutoff <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="co"># return the results on all pairs</span></span>
<span><span class="co">#   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=03h 17m 21s</span></span>
<span><span class="co">######</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Bug in Signac's p-value (01/2025):</span></span>
<span><span class="co"># the two-sided p-values are evaluated with only one tail of the Gaussian distribution</span></span>
<span><span class="co"># https://github.com/stuart-lab/signac/blob/8ecdde2/R/links.R#L481C29-L481C30</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://stuartlab.org/signac/reference/Links.html" class="external-link">Links</a></span><span class="op">(</span><span class="va">signac_res</span><span class="op">)</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span> <span class="co"># the maximum is smaller than 1</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;  0.0000  0.1167  0.2449  0.2449  0.3719  0.5000</span></span>
<span><span class="co"># manually adjust this</span></span>
<span><span class="fu"><a href="https://stuartlab.org/signac/reference/Links.html" class="external-link">Links</a></span><span class="op">(</span><span class="va">signac_res</span><span class="op">)</span><span class="op">$</span><span class="va">pvalue</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://stuartlab.org/signac/reference/Links.html" class="external-link">Links</a></span><span class="op">(</span><span class="va">signac_res</span><span class="op">)</span><span class="op">$</span><span class="va">pvalue</span></span>
<span></span>
<span><span class="va">signac_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://stuartlab.org/signac/reference/Links.html" class="external-link">Links</a></span><span class="op">(</span><span class="va">signac_res</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># consider the same set of peak-gene pairs as scMultiMap</span></span>
<span><span class="va">scMultiMap_res</span><span class="op">$</span><span class="va">pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">scMultiMap_res</span><span class="op">$</span><span class="va">peak</span>, <span class="va">scMultiMap_res</span><span class="op">$</span><span class="va">gene</span>, sep <span class="op">=</span> <span class="st">':'</span><span class="op">)</span></span>
<span><span class="va">signac_res</span><span class="op">$</span><span class="va">pg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">signac_res</span><span class="op">$</span><span class="va">peak</span>, <span class="va">signac_res</span><span class="op">$</span><span class="va">gene</span>, sep <span class="op">=</span> <span class="st">':'</span><span class="op">)</span></span>
<span><span class="va">signac_res</span> <span class="op">&lt;-</span> <span class="va">signac_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">scMultiMap_res</span><span class="op">$</span><span class="va">pg</span>, <span class="va">signac_res</span><span class="op">$</span><span class="va">pg</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># consistent with evaluating scMultiMap's result, we use BH procedure and set the significance cutoff to be 0.2</span></span>
<span><span class="va">signac_res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">signac_res</span><span class="op">$</span><span class="va">pvalue</span>, method <span class="op">=</span> <span class="st">'BH'</span><span class="op">)</span></span>
<span><span class="va">signac_overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/validate_with_assay.html">validate_with_assay</a></span><span class="op">(</span><span class="va">signac_res</span>, <span class="va">sig_assay_gr</span>, pvar <span class="op">=</span> <span class="st">"padj"</span>, p_cutoff <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "p value from one-sided Fisher exact test: 7.61e-02, #overlap: 59, odds ratio: 1.23"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="scent">SCENT<a class="anchor" aria-label="anchor" href="#scent"></a>
</h4>
<p>We further evaluate SCENT, another method benchmarked in scMultiMap’s
<a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>,
you could try the code below. Please note that it could extremely
computationally intensive due to the use of boostrapping. Running the
codes below took <strong>17 hours, 34 minutes</strong> using
<strong>8</strong> cores of an Intel(R) Xeon(R) Gold 6326 CPU @ 2.90GHz.
The high computational costs were also demonstrated by Figure 1c in
scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Tutorial for running SCENT:</span></span>
<span><span class="co"># https://github.com/immunogenomics/SCENT</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"SCENT"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"immunogenomics/SCENT"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SCENT</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract count matrices</span></span>
<span><span class="va">peak_counts</span> <span class="op">&lt;-</span> <span class="va">pbmc_cd14</span><span class="op">[[</span><span class="st">'peaks_ct'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pairs_df</span><span class="op">$</span><span class="va">peak</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">rna_counts</span> <span class="op">&lt;-</span> <span class="va">pbmc_cd14</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pairs_df</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co"># construct metadata for cell-level covariates</span></span>
<span><span class="va">mdata</span> <span class="op">&lt;-</span> <span class="va">pbmc_cd14</span><span class="op">@</span><span class="va">meta.data</span></span>
<span><span class="va">mdata</span><span class="op">$</span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mdata</span><span class="op">)</span></span>
<span><span class="va">mdata</span><span class="op">$</span><span class="va">log_total_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">pbmc_cd14</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">covariates_set</span> <span class="op">&lt;-</span> <span class="st">"log_total_counts"</span></span>
<span></span>
<span><span class="va">SCENT_obj</span> <span class="op">&lt;-</span> <span class="fu">CreateSCENTObj</span><span class="op">(</span>rna <span class="op">=</span> <span class="va">rna_counts</span>, atac <span class="op">=</span> <span class="va">peak_counts</span>,</span>
<span>                            meta.data <span class="op">=</span> <span class="va">mdata</span>,</span>
<span>                            peak.info <span class="op">=</span> <span class="va">pairs_df</span>,</span>
<span>                            covariates <span class="op">=</span> <span class="va">covariates_set</span>,</span>
<span>                            celltypes <span class="op">=</span> <span class="st">'predicted.id'</span><span class="op">)</span> <span class="co"># column name for the cell type label in `mdata`</span></span>
<span></span>
<span><span class="va">start_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">SCENT_obj</span> <span class="op">&lt;-</span> <span class="fu">SCENT_algorithm</span><span class="op">(</span>object <span class="op">=</span> <span class="va">SCENT_obj</span>, celltype <span class="op">=</span> <span class="st">'CD14 Mono'</span>,</span>
<span>                            ncores<span class="op">=</span><span class="fl">8</span>, <span class="co"># for parallelism</span></span>
<span>                             regr <span class="op">=</span> <span class="st">'poisson'</span>, bin <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># default</span></span>
<span><span class="va">end_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SCENT.result</span> <span class="op">&lt;-</span> <span class="va">SCENT_obj</span><span class="op">@</span><span class="va">SCENT.result</span></span>
<span></span>
<span><span class="co"># print elapsed time</span></span>
<span><span class="va">elapsed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/difftime.html" class="external-link">difftime</a></span><span class="op">(</span><span class="va">end_time</span>, <span class="va">start_time</span>, units <span class="op">=</span> <span class="st">"secs"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"scMultiMap elapsed time: %02d:%02d (mm:ss)\n"</span>,</span>
<span>              <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">elapsed</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">3600</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%/%</a></span> <span class="fl">60</span>, <span class="co"># Minutes</span></span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">elapsed</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">60</span><span class="op">)</span><span class="op">)</span>        <span class="co"># Seconds</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SCENT.result</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">SCENT.result</span><span class="op">$</span><span class="va">boot_basic_p</span>, method <span class="op">=</span> <span class="st">'BH'</span><span class="op">)</span></span>
<span><span class="va">SCENT_overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/validate_with_assay.html">validate_with_assay</a></span><span class="op">(</span><span class="va">SCENT.result</span>, <span class="va">sig_assay_gr</span>, pvar <span class="op">=</span> <span class="st">"padj"</span>, p_cutoff <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "p value from one-sided Fisher exact test: 5.77e-01, #overlap: 149, odds ratio: 0.98"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="summary-of-comparison">Summary of comparison<a class="anchor" aria-label="anchor" href="#summary-of-comparison"></a>
</h4>
<p>In this example, <code>scMultiMap</code> identified 246 pairs
overlapped with PCHiC data on CD14 monocytes, while <code>Signac</code>
and <code>SCENT</code> only identified 59 and 149 pairs, respectively.
Moreover, <code>scMultiMap</code> also gave the highest enrichment (odds
ratio=1.74) compared to <code>Signac</code> (1.23) and
<code>SCENT</code> (0.98). In particular, <code>SCENT</code> has a odds
ratio smaller than 1, potentially due to a large number of false
positive discoveries that are not consistent with PCHiC.</p>
<p>This is one example that demonstrates the improved performance by
<code>scMultiMap</code> compared to existing methods. Please refer to
scMultiMap’s <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>
for more examples, including additional consistency analysis with
orthogonal assays, reproducibility analysis across independent
single-cell multimodal datasets, and systematic evaluation of type I
error and power.</p>
<p>Note that the numerical results (i.e., the numbers of reproduced
pairs and odds ratio) presented here differ slightly from those in the
paper, as the published analysis combined four single-cell Multiome PBMC
datasets to increase power (see Methods in <a href="https://www.biorxiv.org/content/10.1101/2024.09.24.614814v1" class="external-link">manuscript</a>).
However, the overall conclusions regarding the comparison across methods
remain unchanged.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="more-vignettes">More vignettes<a class="anchor" aria-label="anchor" href="#more-vignettes"></a>
</h2>
<p>Please visit vignettes: <a href="disease_control.html">scMultiMap for
disease-control studies</a> and <a href="GWAS.html">scMultiMap for
integrative analysis with GWAS results</a> if you are interested in more
examples of analysis with <code>scMultiMap</code>!</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
